### DISCUSSION TITLE
Jack data race

### DISCUSSION THREAD

**Discussion by Andzelmas** [jackaudio/jack2#902] [Q&A] [answered]
Hi,
I might be doing something incorrectly but:
I activate two jack clients, one with output port, the other with input and output. 
First client does some processing and simply memcpy to the out buffer.
The second client memcpy from input buffer to the output buffer.
When the first client out port is connected with input of the second client input port, the ports are of the same address now (not sure which one changes, but if not connected they are different addresses).
This is a data race, no?
Gcc reports it as so, with -fsanitize=thread too.
Is this correct behavior and i should do something more complex with the port buffers if i know they might be connected?
Thank you!

**Comment by kmatheussen** [comment]
I run jack with -fsanitize=thread quite often and can't remember seeing
something like this.
Which version of jack do you use?
Any code you can show?

**Comment by kmatheussen** [comment]
Sorry, I didn't read your comment. You have two clients you transfer data
between inside the same process? I haven't tried that, sorry.

**Reply by Andzelmas** [reply]
If by same process you mean same callback?
Each client has a different callback that processes it on different threads.
Later i'll make a very simplified version of what i have, that produces the same result.

**Reply by Andzelmas** [reply]
Oh, i see what you mean by the same process.
So same process should only have a single client with one callback? 
Wouldn't it be more efficient to have several clients inside your program for different contexts (for lets say a plugin host part, for some sort of audio recorder etc.)?

**Comment by kmatheussen** [comment]
Sorry, I didn't really mean anything. :-) I just wanted to point out
that I don't think I've tried what I think is your scenario while
running thread sanitizer.

Regarding running several clients with several callbacks in the same
process, I don't think there's any reason why that wouldn't work, so
if you get thread sanitizer hits that's probably a bug in jack. But we
need to see your code. :-)

**Reply by Andzelmas** [reply]
Thank you for your help kmatheussen! Really appreciate you taking the time to help me to figure this out.
I've attached a very simplified version of what i'm doing. I build it with gcc -g -fsanitize=thread -x c -o test_jack_mult main.c -ljack
Just connect the b_out to a_in and you will get a data race. The addresses of the ports become the same.
[main.txt](https://github.com/jackaudio/jack2/files/10039436/main.txt)

**Comment by falkTX** [comment]
are you using pipewire by any chance?

reusing the same buffer from one output port as the other input port is quite normal in jack2.
when client A connects to client B, the buffers from client B come directly from audio processed by client A.

there should be no data race though, because when clients are connected to each other they run sequentially. and when not connected to each other they wont share any buffers.

**Reply by Andzelmas** [reply]
I'm running jack simply with jack_control start. 
I have pipewire installed but only because i am using plasma desktop and it has pipewire dependencies. I am not running it as
far as i can tell. I mean i am not using pipewire-jack for jack support or similar.
falkTX do you not get the data race in the example i attached?
Thank you all for help.

**Comment by kmatheussen** [comment, marked as answer]
No errors here:

***@***.***:~/test$ jackd --version
jackdmp version 1.9.12 tmpdir /dev/shm protocol 8
***@***.***:~/test$ gcc -fsanitize=thread jacktest.c -ljack
***@***.***:~/test$ ./a.out
^C
***@***.***:~/test$

**Comment by kmatheussen** [comment]
And jackd output:

***@***.***:~$ jackd -ddummy
jackdmp 1.9.12
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2017 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
no message buffer overruns
no message buffer overruns
no message buffer overruns
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
JackEngine::XRun: client = test_B was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
JackEngine::XRun: client = test_B was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
Cannot write socket fd = 13 err = Broken pipe
CheckRes error
Could not write notification
ClientNotify fails name = test_A notification = 1 val1 = 0 val2 = 0
JackEngine::XRun: client = test_B was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
JackEngine::XRun: client = test_B was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
Cannot write socket fd = 13 err = Broken pipe
CheckRes error
Could not write notification
ClientNotify fails name = test_A notification = 1 val1 = 0 val2 = 0

**Comment by kmatheussen** [comment]
And for alsa:

***@***.***:~$ jackd -dalsa
jackdmp 1.9.12
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2017 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
no message buffer overruns
no message buffer overruns
no message buffer overruns
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
audio_reservation_init
Acquire audio card Audio0
creating alsa driver ... hw:0|hw:0|1024|2|48000|0|0|nomon|swmeter|-|32bit
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for capture: 32bit integer little-endian
ALSA: use 2 periods for capture
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback

**Comment by Andzelmas** [comment]
Well, then i guess this is answered for me. Its not an issue with jack but with my system configuration. Will have to look into that.
Thanks for your help!
By the way this is my output of running the test code:

WARNING: ThreadSanitizer: data race (pid=2655)
  Read of size 8 at 0x7f2b6552fe80 by thread T4:
    #0 memcpy /usr/src/debug/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:827 (libtsan.so.2+0x66a34)
    #1 process_A_rt jack_mult_process_test/main.c:14 (test_jack_mult+0x12bf)
    #2 <null> <null> (libjack.so.0+0x1467e)

  Previous write of size 8 at 0x7f2b6552fe80 by thread T5:
    #0 memset /usr/src/debug/gcc/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:799 (libtsan.so.2+0x652d2)
    #1 process_B_rt jack_mult_process_test/main.c:20 (test_jack_mult+0x1325)
    #2 <null> <null> (libjack.so.0+0x1467e)

  Location is global '<null>' at 0x000000000000 (jack-1000-0+0x2491e80)

  Thread T4 (tid=2665, running) created by main thread at:
    #0 pthread_create /usr/src/debug/gcc/libsanitizer/tsan/tsan_interceptors_posix.cpp:1001 (libtsan.so.2+0x670c9)
    #1 <null> <null> (libjack.so.0+0x3207b)
    #2 <null> <null> (libc.so.6+0x2328f)

  Thread T5 (tid=2666, running) created by main thread at:
    #0 pthread_create /usr/src/debug/gcc/libsanitizer/tsan/tsan_interceptors_posix.cpp:1001 (libtsan.so.2+0x670c9)
    #1 <null> <null> (libjack.so.0+0x3207b)
    #2 <null> <null> (libc.so.6+0x2328f)

-------------------------------------------------------------------
And output of jack -dalsa
~]$ jackd -dalsa
jackdmp 1.9.21
Copyright 2001-2005 Paul Davis and others.
Copyright 2004-2016 Grame.
Copyright 2016-2022 Filipe Coelho.
jackdmp comes with ABSOLUTELY NO WARRANTY
This is free software, and you are welcome to redistribute it
under certain conditions; see the file COPYING for details
no message buffer overruns
no message buffer overruns
no message buffer overruns
JACK server starting in realtime mode with priority 10
self-connect-mode is "Don't restrict self connect requests"
audio_reservation_init
Acquire audio card Audio0
creating alsa driver ... hw:0|hw:0|1024|2|48000|0|0|nomon|swmeter|-|32bit
configuring for 48000Hz, period = 1024 frames (21.3 ms), buffer = 2 periods
ALSA: final selected sample format for capture: 32bit integer little-endian
ALSA: use 2 periods for capture
ALSA: final selected sample format for playback: 32bit integer little-endian
ALSA: use 2 periods for playback
JackEngine::XRun: client = test_A was not finished, state = Running
JackAudioDriver::ProcessGraphAsyncMaster: Process error
JackEngine::XRun: client = test_A was not finished, state = Triggered
JackAudioDriver::ProcessGraphAsyncMaster: Process error
